### PR TITLE
Orchestration/benchmarking review

### PR DESCRIPTION
**Pull request by mloning**
I've had a go at the orchestration functionality again as I want to use this now for a few applications, you can find a list of the things I noticed and that I've changed below, feel free to review the code, make comments, etc. 

Prioritisation:

## Must have
 - [x] udpate example notebook
- [x] write `PairwiseMetric` and `CompositeMetric` wrapper classes
- [x]  given that we also may want to evaluate prediction on the training set and different cv folds, I'm not sure how to best incorporate this into the `Evaluator`/`MetricCalculator` (or `AnalyseResults`/`Losses`) logic, currently not supported, I've added it simply as kwargs, but no automatic aggregation over cv folds is available at the moment as far as I know,
- [x]  No functionality for skipping already fitted strategies or already predicted values, need to implement checks if file already exist and option to not overwrite them, add kwarg for whether or not to overwrite existing files, e.g. `overwrite_fitted_strategies` ,
- [x]  When I specify the same results path for the fitted strategy and predictions, they seem to end up in different folders, I think we should delegate the saving of fitted strategies to the results object for consistency (which then calls the strategy's `save` method), 
- [x]  Add separate `fit` and `predict` methods in addition to `fit_predict()` (or `run`) instead of having `predict_on_runtime` kwarg using `_iter()` method to factor out for loops 
- [x]  I would not convert predicted values to lists to process further in results (`ResultWrapper`) object, but keep them in numpy arrays, unnecessary back and forth conversions as metric calculations make use of arrays again
 - [x]  I'd suggest to implement results loading functionality as iterators to avoid loading in all results at once
 - [x]  No prior checks on validity of input args in orchestration
 - [x]  No master file is created (i.e. the results object is not saved) in case one wants to load all results and
  do the analysis later (may require careful updating if file already exists)
- [x]  Factor out common methods into base classes (with or without using abstract base classes)
- [x]  all results classes have a certain internal logic for where/how to store results, whether inside the object (in-memory) or in a folder/file system on the disk/server, we should factor out this logic into separate methods, as this is needed in multiple cases, e.g. saving, loading and checking if results already exist, in addition, they must keep track of the dataset-strategy combinations that they already seen (something like a registry),
- [x]  `load_dataset()` in `DatasetLoadFromDir` doesn't "load" the data, it creates a list of pointers to the data on the disk, suggest to create a simple function `make_datasets()` to replace the `DatasetLoadFromDir`
 - [x]  Should we not use Bessel's correction (i.e. n - 1 instead of n) when computing sample standard error of the mean?
 - [x]  I'm not sure I can follow the logic of `Losses` (or `MetricCalculator`), why does it make sense to separate that from the `AnalyseResults` class? They seem to operate on two dictionaries, but one of them seems to be a subset of the other, maybe I'm missing something? I'd suggest to store raw metrics in the evaluator and introduce checks for methods that rely on these raw metrics via something like a `check_is_evaluated` (similar to `check_is_fitted`), this is also related to wrapping the metrics into classes
- [x] clone strategies, to avoid updating already fitted strategies
- [x] basic user feedback via printing
 - [x] more extensive unit tests

## Should have
- [ ] add unit tests for `evaluator` methods
- [ ] update all methods on evaluator to work on new internal data representation, also see  https://www.statsmodels.org/stable/stats.html for some additional test implementations, e.g. the sign test, to improve readability, so that we can deprecate `_get_metrics_per_estimator_dataset` and `_get_metrics_per_estimator` methods
- [ ] for saving results inside the `orchestrator` and for loading results in results classes use `_ResultsWrapper` to simply/unify interface, `_ResultsWrapper` needs to have slots for at least: y_true, y_pred, y_proba, index, fit_time, predict_time, strategy_name, dataset_name, cv_fold, train_or_test
- [ ]  No timing of fit and predict available, see https://docs.python.org/3/library/time.html#time.perf_counter, potentially have new `save_timings` and `load_timings` method
- [ ] `orchestrator` cannot make probabilistic predictions, orchestrator tries to make probabilistic predictions using `predict_proba`, but (i) this will only works for some but not all classifiers and it won't work in regression, (ii) strategies currently don't even have a `predict_proba` (not even `TSCStrategy`), and (iii) current computation of `y_proba` fails if `y_pred` contains strings instead of integers which however is an accepted output format for classification I believe, add `predict_proba` to `TSCStrategy`
- [ ] handling of probabilistic metrics in `evaluator`
- [ ] no longer sure that saving results object as a master file is a good idea, as it may cause problems when multiple processes try to update it and because it needs to reflect the state of the directory somehow, maybe better to have a method on results object that allow to infer datasets, strategies and so on, something like a `register_results` method, instead of loading a fully specified dumped result object 
- [ ] separate `predict` method on `orchestrator` which loads and uses already fitted strategies
- [ ] fix UEA results class  

## Could have
- [ ] add `random_state` as input arg to orchestrator which is propagated to all strategies and cv 
- [ ]  perhaps also useful to catch exceptions and skip over them in `orchestrator` instead of breaking execution?
- [ ]  currently only works for ts data input format, add other use cases
- [ ]  better user feedback, logging, keeping track of progress 
- [ ] many docstrings still missing or outdated
- [ ] perhaps metrics shouldn't be wrapped in classes and the evaluator should take care of it internally, working with kwargs (e.g. `pointwise=True`)
- [ ] handling of multiple metrics in `evaluator`

#### Code structure, style and naming
I also change a few names which I find much more intuitive and indicative of what the classes/modules do, but feel to suggest better names!
- [x] rename the module from `experiments` to `benchmarking` (that seems most intuitive to me)
- [x] rename `losses.py` to `evaluation.py` (or something more indicative of what it is used for)
- [x] rename `scores.py` to `metrics.py` as not all metrics are scores, should we follow the sklearn distinction between loss (lower is better) and score (higher is better)?
- [x] rename `AnalyseResults` to `Evaluator` (or maybe `ResultsAnalyser`?) to have a noun for an object
- [x] rename `Result` to `ResultWrapper` as that's its only purpose 
- [x] rename `Losses` to `MetricsCalculator` , and move into `evaluation` file
- [x] only use absolute imports

### CODE REVIEW

**Comment by fkiraly** [comment]
Generally, seems sensible, but below some comments before our discussion tomorrow.

some questions:
- in `UEADataset.load`, why is there a separation of *variables* by train/test index? Where is this needed later, as indicated in the comments?
- orchestrator implementation seems a bit sequential - have you thought how easy this would be to parallelize?

some suggestions:
- there's a case to be made to model "dataset collections" (e.g., all UEA data), or "task collections" (e.g., all UEA data, with tasks to predict target attached to them)
- I still think it might be nice to have separate "set" or "add" methods for the contents of the orchestrator, see e.g. here
https://github.com/alan-turing-institute/mlaut/issues/28
but that's really up for consideration
- for metrics, instead of always making a "func" method/variable crossover, you can simply make the method callable, by using the `__call__` method; I think it's also usually preferable (computationally) to just have pairwise losses already defined in a vectorized fashion, rather than applying a function to all vector entries
- for the "loss functions" design, I think Jackknife is a sensible default theoretically (though fails for quantiles etc), but it might not be computationally. Note that it includes a computational overhead multiplier of "# data points times" since it's calling the recursion that many times.
Would it instead be better to add a "compute standard error" flag instead to avoid computational proliferation by default?
- an alternative design that I've used in R is *wrapping in* the standard error computation. If a metric is an object, you can add a standard error in by wrapping it in a "add the jackknife estimate" (or any other blackbox s.e. estimation method) metric through passing it to the wrapper's constructor. It's maybe a matter of taste, and composability is not really helpful here (who wants the s.e. of a s.e. estimate) but it modularizes out the different s.e. estimation strategies.
- also, it might be helpful to add traits about "higher = better (or not)", the type of metric (input types: continuous? categorical?)
- data I/O design: would it be better to have a dictionary "index" argument (which can be nested) rather than multiple arguments that prescribe which indices you use (e.g., strategy_name, dataset_name, etc)
- in the evaluation class, computations and plotting are merged in a way I'd consider slightly Frankenstein. Particularly, the plotting includes two different sub-tasks: computation of quantities (e.g., performances, CI) and their presentation. The "best practice" pattern for such a situation would be a suitable MVC pattern, but even for "print" outputs there's computation vs presentation.

**Comment by fkiraly** [comment]
as per discussion, fine with everything, and we should later have another round of re-design, modularization and discussion/homogenization wih mlaut. 
For this PR, we should resolve conflicts though.
